File: module_4/python/random_node.py
"""
Title: Linked list random node

Problem:
    Given a singly linked list, return a random node's value from the linked
    list. Each node must have the same probability of being chosen.

    Follow up: What if the linked list is extremely large and its length is
    unknown to you? Could you solve this efficiently without using extra space?

Execution: python random_node.py
"""
from random import random
import unittest


class ListNode:
    def __init__(self, x: int) -> None:
        self.val = x
        self.next = None


class LinkedList():
    def __init__(self, head: ListNode) -> None:
        self.head = head

    def get_random(self) -> int:
        cur_node = self.head
        counter = 1
        choice = -1

        while cur_node is not None:
            if random() < 1/counter:
                choice = cur_node.val
            cur_node = cur_node.next
            counter += 1
        return choice

    def check_random_node(self, node_val: int) -> bool:
        return node_val in [1, 2, 3]



class TestRandomNode(unittest.TestCase):
    """Unit test for get_random."""

    def test_1(self):
        head = ListNode(1)
        head.next = ListNode(2)
        head.next.next = ListNode(3)

        ll = LinkedList(head)
        self.assertEqual(ll.check_random_node(ll.get_random()), True)
        print("Explanation: .")


if __name__ == '__main__':
    unittest.main()

File: module_4/java/RandomNode.java
/*
 *   Title: Linked list random node
 *
 *   Problem:
 *   Given a singly linked list, return a random node's value from the linked
 *   list. Each node must have the same probability of being chosen.
 *
 *   Follow up: What if the linked list is extremely large and its length is unknown
 *   to you? Could you solve this efficiently without using extra space?
 *
 *   Execution: javac RandomNode.java && java RandomNode
 */
import java.util.*;


public class RandomNode {
    ListNode head;
    Random random;

    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    public RandomNode(ListNode h) {
        head = h;       
        random = new Random();        
    }
    
    public int getRandom() {
        
        ListNode c = head;
        int r = c.val;
        for(int i=1;c.next != null;i++){
            
            c = c.next;
            if(random.nextInt(i + 1) == i) r = c.val;                        
        }
        
        return r;
    }
    public static boolean checkRandomNodeFunction(int random_value) {
        if (random_value == 1 || random_value == 2 || random_value == 3) {
            return true;
        }
        return false;
    }
    

    public static void main(String[] args) {
        // Init a singly linked list [1,2,3].
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        RandomNode r_node = new RandomNode(head);
        
        // getRandom() should return either 1, 2, or 3 randomly. Each element
        // should have equal probability of returning.
        int rand_node = r_node.getRandom();
        System.out.println(rand_node);
        assert checkRandomNodeFunction(rand_node) == true;

        System.out.println("Passed all test cases");
    }
    
}
